From 59f8fa4280fddf9955da57d1002ab5e16b2c8744 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Fri, 24 Aug 2018 13:33:36 -0400
Subject: [PATCH] joyent/node-manta#251 specifying a path multiple times to
 mfind results in crash Reviewed by: Robert Mustacchi <rm@joyent.com> Reviewed
 by: Jared Morrow <jm@joyent.com> Approved by: Jared Morrow <jm@joyent.com>

---
 bin/mfind          | 12 +++++++-----
 test/mfind.test.js | 13 +++++++++++++
 2 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/bin/mfind b/bin/mfind
index 6d27942..1368eae 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -8,6 +8,7 @@ var http = require('http');
 var https = require('https');
 var path = require('path');
 var url = require('url');
+var util = require('util');
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
@@ -153,8 +154,9 @@ function printEntry(opts, obj) {
     });
 
     assert(options.paths.length >= 1, 'no paths specified');
-    options.paths.forEach(function (p) {
-        barrier.start(p);
+    options.paths.forEach(function (p, i) {
+        var id = util.format('%d %s', i, p);
+        barrier.start(id);
         client.ftw(p, options, function (err, res) {
             if (err) {
                 if (err.name === 'InvalidDirectoryError') {
@@ -165,14 +167,14 @@ function printEntry(opts, obj) {
                     if (options.limit && totalEntries >= options.limit) {
                         process.exit(0);
                     }
-                    barrier.done(p);
+                    barrier.done(id);
                     return;
                 } else if (err.name === 'NotFoundError') {
                     if (err.path === undefined)
                         err.path = p;
                     printError(err.path, err);
                     lastError = err;
-                    barrier.done(p);
+                    barrier.done(id);
                     return;
                 } else {
                     var epath = p;
@@ -184,7 +186,7 @@ function printEntry(opts, obj) {
             }
 
             res.on('entry', print);
-            res.on('end', barrier.done.bind(barrier, p));
+            res.on('end', barrier.done.bind(barrier, id));
         });
     });
 })();
diff --git a/test/mfind.test.js b/test/mfind.test.js
index af573f4..209c5af 100644
--- a/test/mfind.test.js
+++ b/test/mfind.test.js
@@ -136,6 +136,19 @@ test('mfind TESTDIR', function (t) {
     });
 });
 
+/*
+ * joyent/node-manta#251 specifying a path multiple times to mfind results in
+ * crash.
+ */
+test('mfind TESTDIR TESTDIR (same argument multiple times)', function (t) {
+    forkExecWait({
+        argv: [MFIND, TESTDIR, TESTDIR]
+    }, function (err, info) {
+        t.ifError(err, err);
+        t.done();
+    });
+});
+
 test('mfind -j TESTDIR', function (t) {
     forkExecWait({
         argv: [MFIND, '-j', TESTDIR]
-- 
2.21.0

